Adding batch size argument for inference to fix memory error (#58)

l-gorman opened the pull request:
Changes designed to address issue #58. 

When running the `scripts/train_model.py` script, I was getting an out of memory error at the point of inference. If I understand correctly, all patches were being passed to the model at once. Now we can make these predictions for smaller batches. (not sure if the terminology is right here but this seems to work!)

Please let me know if there's anything else I can do, I know we discussed some different ways of passing batch size argument.

richard-lane commented on fishjaw/model/model.py:
```diff
@@ -431,7 +432,19 @@ def _predict_patches(
     locations = torch.stack(locations)
 
-    return net(tensors).to("cpu").detach(), locations
+    print(tensors.shape)
```
accidental print statement left here

richard-lane reviewed (changes requested):
looks good, assuming it works. there's a hanging print statement - also have you run pylint (it might complain about too many args - you can decide whether to address this or not)?

richard-lane commented on scripts/train_model.py:
```diff
@@ -104,6 +104,8 @@ def main():
         config, data_config, output_dir
     )
 
+    torch.cuda.empty_cache()
```
i think we can get rid of this too

richard-lane commented:
another change - we'll need to change all the other invocations of `plot_inference`, `predict`, etc. to use the new batch_size argument

```
************* Module fishjaw.model.model
fishjaw/model/model.py:450:0: R0913: Too many arguments (6/5) (too-many-arguments)
************* Module fishjaw.visualisation.images_3d
fishjaw/visualisation/images_3d.py:56:0: R0913: Too many arguments (6/5) (too-many-arguments)
************* Module scripts.ablate_attention
scripts/ablate_attention.py:148:17: E1125: Missing mandatory keyword argument 'batch_size' in function call (missing-kwoa)
************* Module scripts.explore_hyperparams
scripts/explore_hyperparams.py:237:18: E1125: Missing mandatory keyword argument 'batch_size' in function call (missing-kwoa)
scripts/explore_hyperparams.py:260:25: E1125: Missing mandatory keyword argument 'batch_size' in function call (missing-kwoa)
************* Module scripts.inference_example
scripts/inference_example.py:170:17: E1125: Missing mandatory keyword argument 'batch_size' in function call (missing-kwoa)
```

richard-lane commented on fishjaw/model/model.py:
```diff
@@ -431,7 +432,19 @@ def _predict_patches(
+        with torch.no_grad():
+            prediction = net(batch).to("cpu").detach()
+        predictions.append(prediction)
+        torch.cuda.empty_cache()  # Clear CUDA cache to free up memory
```
is this needed?

l-gorman replied:
is which part of this needed?

richard-lane replied:
torch.cuda.empty_cache()

richard-lane commented on fishjaw/model/model.py:
```diff
@@ -450,6 +464,7 @@ def predict(
     :param patch_size: the size of the patches to use
     :param patch_overlap: the overlap between patches. Uses a hann window
     :param activation: the activation function to use
+    :param batch_size: the batch size to use
```
id consider changing this to something like "maximum number of patches that will be simultaneously sent to the model", because usually we only talk about batch size when were training

l-gorman commented:
Getting too many arguments error from pylint for the the `plot_inference` function. It seems that `activation`, `patch_size`, and `batch_size` args are all derived from config. Would you prefer me to:

1. simply pass the config to the function and extract vars as needed
2. set batch size to 1 in `_predict_patches` so the argument doesn't need to be included at all
3. ignore pylint and accept that 6 arguments will not lead to world destruction

richard-lane commented:
i think we can just ignore pylint there - if we just pass the whole config it becomes a bit unclear which bits of the config we depend on, it's better to explicitly pass the bits that are needed

the other pylint errors (Missing mandatory keyword argument) do need to be fixed though


PS "too many arguments" is technically not an error, its a refactor suggestion; from the man page:

```
       There are 5 kind of message types :
           * (C) convention, for programming standard violation
           * (R) refactor, for bad code smell
           * (W) warning, for python specific problems
           * (E) error, for probable bugs in the code
           * (F) fatal, if an error occurred which prevented pylint from doing
```

l-gorman commented:
Pushed some changes. I've added a default batch size of 1, but can remove this if needed. I think I have addressed all comments except for [this](https://github.com/JGIBristol/zebrafish_jaw_segmentation/pull/59#discussion_r1906906720). `train_model.py` script still seems to work with these changes. Testing with more epochs now to see if the outputs look approximately right.
